satellite/satellitedb/bandwidthagreement.go

```diff
@@ -5,74 +5,99 @@ package satellitedb
+func (b *bandwidthagreement) GetUplinkStats(ctx context.Context, from, to time.Time) (stats []bwagreement.UplinkStat, err error) {
+
+	var uplinkSQL = fmt.Sprintf(`SELECT uplink_id, SUM(total), 
+		COUNT(CASE WHEN action = %d THEN total ELSE null END), 
+		COUNT(CASE WHEN action = %d THEN total ELSE null END), COUNT(*)
+		FROM bwagreements WHERE created_at > ? 
+		AND created_at <= ? GROUP BY uplink_id ORDER BY uplink_id`,
+		pb.BandwidthAction_PUT, pb.BandwidthAction_GET)
+	rows, err := b.db.DB.Query(b.db.Rebind(uplinkSQL), from, to)
 	if err != nil {
 		return nil, err
 	}
-	agreements := make([]bwagreement.Agreement, len(rows))
-	for i, entry := range rows {
-		rba := pb.RenterBandwidthAllocation{}
-		err := proto.Unmarshal(entry.Data, &rba)
+	defer func() { err = errs.Combine(err, rows.Close()) }()
+	for rows.Next() {
+		var nodeID []byte
+		stat := bwagreement.UplinkStat{}
+		err := rows.Scan(&nodeID, &stat.TotalBytes, &stat.PutActionCount, &stat.GetActionCount, &stat.TotalTransactions)
 		if err != nil {
-			return nil, err
+			return stats, err
 		}
-		agreement := &agreements[i]
-		agreement.Agreement = rba
-		agreement.CreatedAt = entry.CreatedAt
+		id, err := storj.NodeIDFromBytes(nodeID)
+		if err != nil {
+			return stats, err
+		}
+		stat.NodeID = id
+		stats = append(stats, stat)
 	}
-	return agreements, nil
+	return stats, nil
 }
 
-func (b *bandwidthagreement) GetAgreementsSince(ctx context.Context, since time.Time) ([]bwagreement.Agreement, error) {
-	rows, err := b.db.All_Bwagreement_By_CreatedAt_Greater(ctx, dbx.Bwagreement_CreatedAt(since))
+//GetTotals returns the sum of each bandwidth type after (exluding) a given date range
+func (b *bandwidthagreement) GetTotals(ctx context.Context, from, to time.Time) (bwa map[storj.NodeID][]int64, err error) {
+	var getTotalsSQL = fmt.Sprintf(`SELECT storage_node_id, 
+		SUM(CASE WHEN action = %d THEN total ELSE 0 END),
+		SUM(CASE WHEN action = %d THEN total ELSE 0 END), 
+		SUM(CASE WHEN action = %d THEN total ELSE 0 END),
+		SUM(CASE WHEN action = %d THEN total ELSE 0 END), 
+		SUM(CASE WHEN action = %d THEN total ELSE 0 END)
+		FROM bwagreements WHERE created_at > ? AND created_at <= ? 
+		GROUP BY storage_node_id ORDER BY storage_node_id`, pb.BandwidthAction_PUT,
+		pb.BandwidthAction_GET, pb.BandwidthAction_GET_AUDIT,
+		pb.BandwidthAction_GET_REPAIR, pb.BandwidthAction_PUT_REPAIR)
+	rows, err := b.db.DB.Query(b.db.Rebind(getTotalsSQL), from, to)
 	if err != nil {
 		return nil, err
 	}
+	defer func() { err = errs.Combine(err, rows.Close()) }()
 
-	agreements := make([]bwagreement.Agreement, len(rows))
-	for i, entry := range rows {
-		rba := pb.RenterBandwidthAllocation{}
-		err := proto.Unmarshal(entry.Data, &rba)
+	totals := make(map[storj.NodeID][]int64)
+	for i := 0; rows.Next(); i++ {
+		var nodeID []byte
+		data := make([]int64, len(pb.BandwidthAction_value))
+		err := rows.Scan(&nodeID, &data[0], &data[1], &data[2], &data[3], &data[4])
```

Comment:
Can we use here already the actual types, arrays of unknown index are confusing.